### PR TITLE
21 random

### LINKED ISSUE #21: make random controller for items, invoice items, invoices

### PR DESCRIPTION
**Pull request by Laszlo-JFLMTCO**
@akintner This PR closes #21 . It also has all changes from PR #47 and from PR #48 in addition to all of your (Person A) `find` and `find_all` endpoints.

I took a different approach for the `random` functionality implementation: instead of using `take` method within the controller, I added a model level `random` method, what is using `all.shuffle.first` that is returning randomized single object vs `take` returning the same single object. Let's discuss any questions you might have and agree on how to standardize the `random` implementation later this morning.

FYI: @case-eee

### CODE REVIEW

**Comment by akintner** [comment]
I do like this better, as the randomness of the model method is more legitimately random than what I did. Nice, let's keep this and I'll change what I did last night to match.

File: app/models/invoice.rb [hunk 1]
```diff
@@ -1,2 +1,42 @@
 class Invoice < ApplicationRecord
+
+  def self.find_by_params(param)
```
Comment (case-eee) [inline app/models/invoice.rb:3]: @Laszlo-JFLMTCO @akintner i believe if you use strong params in your controller, you don't need to write this method. you could just pass your params to the `find_by` method that's coming from ActiveRecord

File: app/models/invoice.rb [hunk 2]
```diff
@@ -1,2 +1,42 @@
...
+    elsif param["created_at"]
+      where(created_at: param["created_at"])
+    elsif param["updated_at"]
+      where(updated_at: param["updated_at"])
+    else
+      nil
+    end
+  end
+
+  def self.random
```
Comment (case-eee) [inline app/models/invoice.rb:39]: you could also check into ordering by random 👍 meaning use the `order` method and pass it something that's asking for a random order

@Laszlo-JFLMTCO @akintner

File: app/models/invoice_item.rb [hunk 3]
```diff
@@ -1,2 +1,46 @@
 class InvoiceItem < ApplicationRecord
+  def self.find_by_params(param)
```
Comment (case-eee) [inline app/models/invoice_item.rb:2]: same as above!

File: spec/requests/api/v1/invoice_items_find_all_spec.rb [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    test_invoice_items[2].update(item_id: 1)
+    get "/api/v1/invoice_items/find_all?item_id=#{test_invoice_items.first['item_id']}"
+
+    invoice_items = JSON.parse(response.body)
+    invoice_item = invoice_items.first
+
+    expect(response).to be_success
+    expect(invoice_items.count).to eq(2)
+    expect(invoice_item).to be_a(Hash)
+    expect(invoice_item).to have_key('id')
```
Comment (case-eee) [inline spec/requests/api/v1/invoice_items_find_all_spec.rb:41]: if we're going to check each individual key and value, why not just check to see if the json parsed is equivalent to what you expect

**Review by case-eee — COMMENTED** [review]
@Laszlo-JFLMTCO @akintner left you some comments. let me know if you have questions 👍

File: spec/requests/api/v1/invoice_items_find_all_spec.rb (hunk 4 above)
Comment (Laszlo-JFLMTCO) [reply spec/requests/api/v1/invoice_items_find_all_spec.rb:41]: @case-eee The thinking behind not comparing with actual JSON response, but checking each key and their class individually is that we had to go through several iterations of content formatting updates. In this case we can confirm the correct keys are coming through, but accurate content being returned we can make check at `model` level testing.

FYI @akintner

Comment (case-eee) [reply spec/requests/api/v1/invoice_items_find_all_spec.rb:41]: @Laszlo-JFLMTCO sounds good to me!! thanks for explaining 👍

**Pull request merged**